diff --git a/CMakeLists.txt b/CMakeLists.txt
index a1937f1..85b03f2 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -20,7 +20,7 @@
 
 cmake_minimum_required(VERSION 3.2 FATAL_ERROR)
 # The version of CMake which is not compatible with the old CUDA CMake commands.
-set(CMAKE_VERSION_THRESHOLD "3.10.0")
+set(CMAKE_VERSION_THRESHOLD "3.15.0")
 
 if(${CMAKE_VERSION} VERSION_LESS ${CMAKE_VERSION_THRESHOLD})
   project(onnx2trt LANGUAGES CXX C)
